File: src/studica_control/src/components/examples/python/titan_example.py
#!/usr/bin/env python3
"""Titan motor controller — command motors via topics and read encoder/RPM feedback.

Before the run loop: Titan2 setup via titan_cmd (PID type, autotune, target velocity).
Then runs motor 0 at target RPM for 3 seconds, stops for 2 seconds,
then resets the encoder via the service.

Run:  ros2 run studica_control titan_example.py
Requires: studica_launch.py running, titan enabled in params.yaml
          sensor name "titan0", m_0 encoder_mode: quadrature

Topics published:   /titan0/m_N/cmd      (Float64)  duty cycle -1.0 to 1.0
Topics subscribed:  /titan0/m_0/encoder, /titan0/m_0/rpm
Service:            /titan0/titan_cmd    (SetData)
"""
import time

import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64
from studica_control.srv import SetData

SENSOR = 'titan0'

PID_MCV2 = 2
MOTOR = 0
TARGET_RPM = 10.0
AUTOTUNE_WAIT_S = 15


class TitanExample(Node):
    def __init__(self):
        super().__init__('titan_example')

        self.cmd_pubs = [
            self.create_publisher(Float64, f'/{SENSOR}/m_{i}/cmd', 10)
            for i in range(4)
        ]

        self.create_subscription(Float64, f'/{SENSOR}/m_0/encoder', self.on_encoder, 10)
        self.create_subscription(Float64, f'/{SENSOR}/m_0/rpm', self.on_rpm, 10)

        self.client = self.create_client(SetData, f'/{SENSOR}/titan_cmd')
        self.get_logger().info(f'Titan example ready. Sensor: {SENSOR}')

    def on_encoder(self, msg):
        self.get_logger().info(f'm_0 encoder distance: {msg.data:.4f}')

    def on_rpm(self, msg):
        self.get_logger().info(f'm_0 rpm: {msg.data:.2f}')

    def set_speed(self, motor: int, duty: float):
        msg = Float64()
        msg.data = float(duty)
        self.cmd_pubs[motor].publish(msg)

    def call_service(self, command: str, motor: int = 0, speed: float = 0.0,
                     int_value: int = 0, dist_per_tick: float = 0.0):
        if not self.client.wait_for_service(timeout_sec=5.0):
            self.get_logger().error('titan_cmd service not available')
            return None
        req = SetData.Request()
        req.params = command
        req.initparams.n_encoder = motor
        req.initparams.speed = float(speed)
        req.initparams.int_value = int_value
        req.initparams.dist_per_tick = dist_per_tick
        future = self.client.call_async(req)
        rclpy.spin_until_future_complete(self, future, timeout_sec=5.0)
        result = future.result()
        if result and result.success:
            self.get_logger().info(f'{command}: {result.message}')
        elif result:
            self.get_logger().error(f'{command}: {result.message}')
        else:
            self.get_logger().error(f'{command}: no response')
        return result


def main():
    rclpy.init()
    node = TitanExample()

    if not node.client.wait_for_service(timeout_sec=10.0):
        node.get_logger().error('titan_cmd not available — is studica_launch running?')
        rclpy.shutdown()
        return

    node.get_logger().info('setup: set_pid_type MCV2 on all motors')
    if not node.call_service('set_pid_type', int_value=PID_MCV2):
        rclpy.shutdown()
        return

    node.get_logger().info(f'setup: autotune all ({AUTOTUNE_WAIT_S} s)')
    if not node.call_service('autotune'):
        rclpy.shutdown()
        return

    end = time.time() + AUTOTUNE_WAIT_S
    while time.time() < end:
        rclpy.spin_once(node, timeout_sec=0.05)

    if not node.call_service('set_target_velocity', motor=MOTOR, speed=TARGET_RPM):
        rclpy.shutdown()
        return

    node.get_logger().info(f'running at {TARGET_RPM:.1f} rpm')
    end = time.time() + 3.0
    while time.time() < end:
        rclpy.spin_once(node, timeout_sec=0.05)

    node.call_service('set_target_velocity', motor=MOTOR, speed=0.0)
    node.set_speed(MOTOR, 0.0)
    node.get_logger().info('stopped')

    end = time.time() + 2.0
    while time.time() < end:
        rclpy.spin_once(node, timeout_sec=0.05)

    node.call_service('reset_encoder', motor=MOTOR)
    node.get_logger().info('done — encoder reset')

    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/studica_control/src/components/examples/cpp/titan_example.cpp
/*
 * titan_example.cpp
 *
 * Command motors via topics and read encoder/RPM feedback.
 * Before the run loop: Titan2 setup via titan_cmd (PID type, autotune, target velocity).
 * Then runs motor 0 at target RPM for 3 seconds, stops for 2 seconds
 * then resets the encoder via the service.
 *
 * Run:      ros2 run studica_control titan_example
 * Requires: studica_launch.py running, titan enabled in params.yaml
 *           sensor name "titan0", m_0 encoder_mode: quadrature
 *
 * Topics published:   /titan0/m_N/cmd  (std_msgs/Float64)  duty cycle -1.0 to 1.0
 * Topics subscribed:  /titan0/m_0/encoder  (std_msgs/Float64)
 *                     /titan0/m_0/rpm      (std_msgs/Float64)
 * Service:            /titan0/titan_cmd    (studica_control/SetData)
 */

#include <array>
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64.hpp"
#include "studica_control/srv/set_data.hpp"

using namespace std::chrono_literals;

static const std::string SENSOR = "titan0";

static constexpr int32_t PID_MCV2 = 2;
static constexpr int MOTOR = 0;
static constexpr float TARGET_RPM = 10.0f;
static constexpr int AUTOTUNE_WAIT_S = 15;

class TitanExample : public rclcpp::Node {
public:
    TitanExample() : Node("titan_example"), phase_(Phase::RUNNING), setup_done_(false) {
        for (int i = 0; i < 4; ++i) {
            cmd_pubs_[i] = create_publisher<std_msgs::msg::Float64>(
                "/" + SENSOR + "/m_" + std::to_string(i) + "/cmd", 10);
        }

        // quadrature mode feedback — comment out if using encoder_mode: absolute
        enc_sub_ = create_subscription<std_msgs::msg::Float64>(
            "/" + SENSOR + "/m_0/encoder", 10,
            [this](std_msgs::msg::Float64::SharedPtr msg) {
                RCLCPP_INFO(get_logger(), "m_0 encoder distance: %.4f", msg->data);
            });
        rpm_sub_ = create_subscription<std_msgs::msg::Float64>(
            "/" + SENSOR + "/m_0/rpm", 10,
            [this](std_msgs::msg::Float64::SharedPtr msg) {
                RCLCPP_INFO(get_logger(), "m_0 rpm: %.2f", msg->data);
            });

        // absolute mode feedback — uncomment if encoder_mode: absolute
        // angle_sub_ = create_subscription<std_msgs::msg::Float64>(
        //     "/" + SENSOR + "/m_0/angle", 10, [this](std_msgs::msg::Float64::SharedPtr msg) {
        //         RCLCPP_INFO(get_logger(), "m_0 angle: %.2f deg", msg->data); });

        client_ = create_client<studica_control::srv::SetData>("/" + SENSOR + "/titan_cmd");

        phase_start_ = now();
        timer_ = create_wall_timer(100ms, std::bind(&TitanExample::tick, this));
        RCLCPP_INFO(get_logger(), "Titan example ready — waiting for titan_cmd, then setup");
    }

private:
    enum class Phase { RUNNING, STOPPED, DONE };

    bool call_service(const std::string& command, int motor = 0, float speed = 0.f, int32_t int_value = 0) {
        if (!client_->wait_for_service(2s)) {
            RCLCPP_ERROR(get_logger(), "titan_cmd service not available");
            return false;
        }
        auto req = std::make_shared<studica_control::srv::SetData::Request>();
        req->params = command;
        req->initparams.n_encoder = motor;
        req->initparams.speed = speed;
        req->initparams.int_value = int_value;

        auto future = client_->async_send_request(req);
        if (rclcpp::spin_until_future_complete(shared_from_this(), future, 5s) !=
            rclcpp::FutureReturnCode::SUCCESS) {
            RCLCPP_ERROR(get_logger(), "%s: timed out", command.c_str());
            return false;
        }
        const auto resp = future.get();
        if (resp->success) {
            RCLCPP_INFO(get_logger(), "%s: %s", command.c_str(), resp->message.c_str());
        } else {
            RCLCPP_ERROR(get_logger(), "%s: %s", command.c_str(), resp->message.c_str());
        }
        return resp->success;
    }

    /** One-time Titan2 setup before the original RUNNING / STOPPED loop. */
    void tick_setup() {
        switch (setup_index_) {
            case 0:
                if (!client_->wait_for_service(0s)) {
                    return;
                }
                RCLCPP_INFO(get_logger(), "setup: set_pid_type MCV2 on all motors");
                if (!call_service("set_pid_type", 0, 0.f, PID_MCV2)) {
                    phase_ = Phase::DONE;
                    return;
                }
                setup_index_ = 1;
                break;
            case 1:
                RCLCPP_INFO(get_logger(), "setup: autotune all (%d s)", AUTOTUNE_WAIT_S);
                if (!call_service("autotune")) {
                    phase_ = Phase::DONE;
                    return;
                }
                setup_index_ = 2;
                phase_start_ = now();
                break;
            case 2:
                if ((now() - phase_start_).seconds() < AUTOTUNE_WAIT_S) {
                    return;
                }
                if (!call_service("set_target_velocity", MOTOR, TARGET_RPM)) {
                    phase_ = Phase::DONE;
                    return;
                }
                setup_done_ = true;
                phase_ = Phase::RUNNING;
                phase_start_ = now();
                RCLCPP_INFO(get_logger(), "running at %.1f rpm", TARGET_RPM);
                break;
            default:
                break;
        }
    }

    void set_speed(int motor, double duty) {
        std_msgs::msg::Float64 msg;
        msg.data = duty;
        cmd_pubs_[motor]->publish(msg);
    }

    void call_reset_encoder(int motor) {
        if (!client_->wait_for_service(2s)) {
            RCLCPP_ERROR(get_logger(), "titan_cmd service not available");
            return;
        }
        auto req = std::make_shared<studica_control::srv::SetData::Request>();
        req->params = "reset_encoder";
        req->initparams.n_encoder = motor;
        client_->async_send_request(req,
            [this](rclcpp::Client<studica_control::srv::SetData>::SharedFuture f) {
                RCLCPP_INFO(get_logger(), "reset_encoder: %s", f.get()->message.c_str());
            });
    }

    void tick() {
        if (phase_ == Phase::DONE) {
            return;
        }

        if (!setup_done_) {
            tick_setup();
            return;
        }

        const double elapsed = (now() - phase_start_).seconds();

        if (phase_ == Phase::RUNNING && elapsed >= 3.0) {
            call_service("set_target_velocity", MOTOR, 0.f);
            set_speed(MOTOR, 0.0);
            phase_ = Phase::STOPPED;
            phase_start_ = now();
            RCLCPP_INFO(get_logger(), "stopped");

        } else if (phase_ == Phase::STOPPED && elapsed >= 2.0) {
            call_reset_encoder(MOTOR);
            phase_ = Phase::DONE;
            timer_->cancel();
            RCLCPP_INFO(get_logger(), "done — encoder reset");
        }
    }

    std::array<rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr, 4> cmd_pubs_;
    rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr enc_sub_;
    rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr rpm_sub_;
    rclcpp::Client<studica_control::srv::SetData>::SharedPtr client_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Time phase_start_;
    Phase phase_;
    bool setup_done_;
    int setup_index_ = 0;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TitanExample>());
    rclcpp::shutdown();
    return 0;
}
